backend/app-orchestrator-service/src/main/kotlin/services/JobResourceService.kt

```diff
@@ -474,6 +518,125 @@ class JobResourceService {
+            jobNotifications.set(new)
+        }
+
+        val newNotifications = jobNotifications.get()
+
+        newNotifications[user]!!.add(
+            JobNotificationInfo(
+                newState,
+                jobId,
+                appTitle,
+                jobSpecification.name,
+            )
+        )
+
+        jobNotifications.set(newNotifications)
+
+
+    }
+
+    private suspend fun sendNotifications() {
+        val handledTypes = mutableListOf<JobState>()
+
+        for (user in jobNotifications.get().keys) {
+            val notifications = jobNotifications.get()[user] ?: continue
+
+            val summarizedNotifications: List<Notification> = notifications.mapNotNull { notification ->
+                if (!handledTypes.contains(notification.type)) {
+                    val sameType = notifications.filter { notification.type == it.type }
+                    handledTypes.add(notification.type)
+
+                    val jobIds = sameType.map { JsonPrimitive(it.jobId) }
+                    val appTitles = sameType.map { JsonPrimitive(it.appTitle) }
+
+                    val type = when (notification.type) {
+                        JobState.SUCCESS -> "JOB_COMPLETED"
+                        JobState.RUNNING -> "JOB_STARTED"
+                        JobState.FAILURE -> "JOB_FAILED"
+                        JobState.EXPIRED -> "JOB_EXPIRED"
+                        else -> return;
+                    }
+
+                    Notification(
+                        type,
+                        "",
+                        meta = JsonObject(
+                            mapOf(
+                                "jobIds" to JsonArray(jobIds),
+                                "appTitles" to JsonArray(appTitles),
+                            )
+                        )
+                    )
+                } else {
+                    null
+                }
+            }
+
+            summarizedNotifications.forEach { notification ->
+                NotificationDescriptions.create.call(
+                    CreateNotification(
+                        user,
+                        notification
+                    ),
+                    serviceClient
+                )
+            }
+
+            handledTypes.clear()
+
+            val summarizedMails = notifications.mapNotNull { notification ->
+                if (!handledTypes.contains(notification.type)) {
+                    val sameType = notifications.filter { notification.type == it.type }
+                    handledTypes.add(notification.type)
+
+                    val jobIds = sameType.map {
+                        if (it.jobName != null) {
+                            "${it.jobName} (${it.jobId})"
+                        } else {
+                            it.jobId
+                        }
+                    }
+
+                    val appTitles = sameType.map { it.appTitle }
+
+                    when (notification.type) {
```

Comment:
I'm probably nitpicking, but I think it would be useful if we could bundle different states into a single email as well.

For example, assume these events occur within the "collection window":

- Job 41231 start
- Job 41232 start
- Job 41233 start
- Job 41231 fail
- Job 41000 completes
- Job 41232 fail

Would result in a single email:

Subject: Your job(s) on UCloud has started
Body:
3 of your jobs have started and 3 of your jobs have completed.

- [LINK TO JOB: 41231]: ${jobTitle ?: applicationTitle} has started
- [LINK TO JOB: 41232]: ${jobTitle ?: applicationTitle} has started
- [LINK TO JOB: 41233]: ${jobTitle ?: applicationTitle} has started
- [LINK TO JOB: 41231]: ${jobTitle ?: applicationTitle} has completed (failure)
- [LINK TO JOB: 41000]: ${jobTitle ?: applicationTitle} has completed
- [LINK TO JOB: 41232]: ${jobTitle ?: applicationTitle} has failed